File: src/main/java/com/ddd/trainings/idea/api/app/TrainingIdeaService.java
package com.ddd.trainings.idea.api.app;

import com.ddd.trainings.idea.domain.IdeaId;
import com.ddd.trainings.idea.domain.TrainingDurationVO;
import com.ddd.trainings.idea.domain.TrainingIdea;
import com.ddd.trainings.idea.infra.TrainingIdeaRepo;
import com.ddd.trainings.proposal.domain.ProposalId;
import com.ddd.trainings.proposal.domain.ReviewerPolicyFactory;
import com.ddd.trainings.proposal.domain.TrainingProposal;
import com.ddd.trainings.proposal.infra.TrainingProposalRepo;
import org.jmolecules.ddd.annotation.Service;

@Service
public class TrainingIdeaService {

  private final TrainingIdeaRepo ideaRepo;
  private final TrainingProposalRepo proposalRepo;
  private final ReviewerPolicyFactory reviewerPolicyFactory;

  public TrainingIdeaService(TrainingIdeaRepo ideaRepo, TrainingProposalRepo proposalRepo, ReviewerPolicyFactory reviewerPolicyFactory) {
    this.ideaRepo = ideaRepo;
    this.proposalRepo = proposalRepo;
    this.reviewerPolicyFactory = reviewerPolicyFactory;
  }

  public IdeaId createTrainingIdea(String trainerId, String name) {
    // TODO: Implement creating idea using IdeaFactory
    // TODO: save it in the repository
    return null;
  }

  public ProposalId proposeIdea(String ideaId) {
    TrainingIdea idea = ideaRepo.findById(IdeaId.from(ideaId))
        .orElseThrow(() -> new RuntimeException("Idea not found"));

    TrainingProposal proposal = idea.propose(reviewerPolicyFactory.byKeywords(idea.getName()));
    proposalRepo.save(proposal);
    return proposal.getId();
  }

  public void editDuration(String ideaId, int days) {
    TrainingIdea idea = ideaRepo.findById(IdeaId.from(ideaId))
        .orElseThrow(() -> new RuntimeException("Idea not found"));
    idea.editDuration(TrainingDurationVO.from(days));
    ideaRepo.save(idea);
  }
}

File: src/Ddd/Trainings/Idea/App/TrainingIdeaService.cs
using Ddd.Trainings.Idea.Domain;
using Ddd.Trainings.Idea.Domain.Ports;
using Ddd.Trainings.Proposal.Domain;
using Ddd.Trainings.Proposal.Domain.Ports;

namespace Ddd.Trainings.Idea.Api.App;

public class TrainingIdeaService
{
    private readonly ITrainingIdeaRepo _ideaRepo;
    private readonly ITrainingProposalRepo _proposalRepo;
    private readonly ReviewerPolicyFactory _reviewerPolicyFactory;

    public TrainingIdeaService(
        ITrainingIdeaRepo ideaRepo,
        ITrainingProposalRepo proposalRepo,
        ReviewerPolicyFactory reviewerPolicyFactory)
    {
        _ideaRepo = ideaRepo;
        _proposalRepo = proposalRepo;
        _reviewerPolicyFactory = reviewerPolicyFactory;
    }

    public IdeaId CreateTrainingIdea(string trainerId, string name)
    {
        TrainingIdea idea = TrainingIdeaFactory.Create(trainerId, name);
        TrainingIdea saved = _ideaRepo.Save(idea);
        return saved.GetId();
    }

    public void EditDuration(string ideaId, int days)
    {
        TrainingIdea idea = _ideaRepo.FindById(IdeaId.From(ideaId)) ?? throw new Exception("Idea not found");
        idea.EditDuration(TrainingDurationVO.From(days));
        _ideaRepo.Save(idea);
    }

    public ProposalId ProposeIdea(string ideaId)
    {
        // TODO: Propose idea and persist updated entity
        return null;
    }
}